File: zig/binary-search/binary_search.zig
pub fn binarySearch(comptime T: type, target: T, items: []const T) ?usize {
    var min: usize = 0;
    if (items.len == 0) return null;
    var max: usize = items.len - 1;
    while (min <= max) {
        const midpoint: usize = (min + max) / 2;

        if (items[midpoint] == target) {
            return midpoint;
        } else if (items[midpoint] > target) {
            if (midpoint == 0) break;
            max = midpoint - 1;
        } else {
            min = midpoint + 1;
        }
    }

    return null;
}

File: c/binary-search/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, size_t length) {
  int min = 0;
  int max = length-1;
  while (min <= max) {
    int mid = (max + min) / 2;
    int pivot = arr[mid];
    if (pivot == value) {
      return &arr[mid];
    } else if (value > pivot) {
      min = mid + 1;
    } else {
      max = mid - 1;
    }
  }
  return NULL;
}
